src/actions.ts

```diff
@@ -302,9 +302,19 @@ export function stop<TContext, TEvent extends EventObject>(
  *
  * @param assignment An object that represents the partial context to update.
  */
-export const assign = <TContext, TEvent extends EventObject = EventObject>(
+export function assign<TContext, TEvent extends EventObject = EventObject>(
+  // trick to make TS inferring TContext based on return type of `assignment` instead of its arguments (microsoft/TypeScript#31933)
+  assignment: TContext extends any ?
+    (Assigner<TContext, TEvent> | PropertyAssigner<TContext, TEvent>)
+    : never
+): AssignAction<TContext, TEvent>;
+export function assign<TContext, TEvent extends EventObject = EventObject>(
+  // second overload to make this work better in generic contexts (where TS has problems with resolving conditional types)
```

Comment:
without this TS was not able to allow this:
https://github.com/davidkpiano/xstate/blob/6eea481d82edf448cec9444ecde467e53d573be6/src/scxml.ts#L77
even though it should - because TContext extends here object and object extends any, but TS can't propagate this information right now through conditional types.